sdk: use graphql endpoint for zklogin verify

joyqvq opened the pull request:
## Description 

add support for zklogin personal msg verify by invoking graphql endpoint. 

## Test plan 

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

joyqvq commented on crates/sui-graphql-rpc/src/types/zklogin_verify_signature.rs:
```diff
@@ -36,7 +36,7 @@ pub(crate) enum ZkLoginIntentScope {
 pub(crate) struct ZkLoginVerifyResult {
     /// The boolean result of the verification. If true, errors should be empty.
-    success: bool,
+    pub success: bool,
```
the typescript cannot find this type definition, i think making this pub should do.

amnn replied:
This shouldn't be the case -- (i.e. I think it should show up in the GraphQL schema even without the `pub` here) -- but it's also fine to have it!

joyqvq commented on sdk/typescript/src/cryptography/signature.ts:
```diff
@@ -72,24 +70,13 @@ export function parseSerializedSignature(serializedSignature: SerializedSignatur
+				signatureScheme: res.signatureScheme,
+				zkLogin: res.zkLogin,
+				signature: res.signature,
+				publicKey: res.publicKey,
```
for some reason if i do 

`return parseSerializedZkLoginSignature(serializedSignature)` here the typescript complains about the type undefined the json. i recasted it here, lmk if this is the right ts approach

joyqvq commented on sdk/typescript/src/verify/index.ts:
```diff
@@ -29,6 +31,20 @@ export async function verifyPersonalMessage(
 ): Promise<PublicKey> {
```
this method is not exposed in the sdk, i added the support here too for completeness.

hayes-mysten commented on sdk/typescript/src/zklogin/publickey.ts:
```diff
@@ -77,3 +84,84 @@ export function toZkLoginPublicIdentifier(
+		url:
+			network === 'testnet'
+				? 'https://sui-testnet.mystenlabs.com/graphql'
+				: 'https://sui-mainnet.mystenlabs.com/graphql',
```
So far we have avoided hardcoding endpoints into the SDK, especially in ways that can't be overwritten.  The GraphQL API is also likely to introduce path-based routing for versioned schemas, which may change these urls.  We probably need to think through the best way to handle this

hayes-mysten commented on sdk/typescript/test/unit/zklogin/signature.test.ts:
```diff
@@ -60,4 +64,16 @@ describe('zkLogin signature', () => {
+		let pk = new ZkLoginPublicIdentifier(parsed.publicKey);
+
+		// returns error bc the test signature is for max epoch 10, should not verify.
+		let res = await pk.verifyPersonalMessage(fromB64(bytes), parsed.signature);
```
this will hit mainnet, does this method work in localnet?  We currently don't have any other tests that make requests to services other than the localnet instance we are running for tests

joyqvq replied:
i dont know if theres a graphql client being spinned up during the test. i think hitting mainnet is ok unless there are other concerns. the endpoint should be stable

hayes-mysten commented on sdk/typescript/test/unit/zklogin/signature.test.ts:
```diff
@@ -60,4 +64,16 @@ describe('zkLogin signature', () => {
+
+		// returns error bc the test signature is for max epoch 10, should not verify.
+		let res = await pk.verifyPersonalMessage(fromB64(bytes), parsed.signature);
+		expect(res).toBe(false);
```
We probably need a test that actually verifies.  I just noticed I don't think the current implementation is correct,  We are passing the digest of the intent message to the GraphQL endpoint, which is not what it's expecting

joyqvq replied:
this will be slightly tricky, because a valid signature has to have a valid max_epoch. but i can potentially set this to a large number like 1000000 and since devnet is wiped every week, it should always stay valid. 

we are not passing the digest to the endpoint - this test is a message "Hello world" encoded in base64 here.

hayes-mysten commented on sdk/typescript/src/zklogin/publickey.ts:
```diff
@@ -75,5 +87,82 @@ export function toZkLoginPublicIdentifier(
+		variables: {
+			bytes,
+			signature,
+			intent_scope: 'PERSONAL_MESSAGE',
```
Should have noticed this earlier, but this won't actually work as is.

The public keys verifyPersonalMessage method will call verifyWithIntent, which then calls verify with the digest of the intent message.  The `verify` method we implemented on zksend public key is intended for both personal messages and transactions, and is passed only the signature, and the digest of the intent message,  rather than the raw message bytes.

joyqvq replied:
graphql always accepts the raw personal message bytes not the digest.

hayes-mysten replied:
The issue is that PublicKeyInitData.verify doesn't get the message bytes, it gets the digest, so while `graphqlVerifyZkLoginSignature` will work, `pk.verifyPersonalMessage` will not

joyqvq commented:
rebased and added tests for a success/fail case for sig verify (by directly implementing verifyPersonalMessage and leaves the verify method throw error) cc @hayes-mysten PTAL 🙏

hayes-mysten commented on sdk/typescript/src/zklogin/publickey.ts:
```diff
@@ -75,5 +96,84 @@ export function toZkLoginPublicIdentifier(
+		variables: {
+			bytes,
+			signature,
+			intent_scope: 'PERSONAL_MESSAGE',
```
Could we pass in the intentScope so that we can implement verifyPersonalMessage and verifyTransactionBlock on the PublicKey class?

joyqvq replied:
good point! revised and added tests

hayes-mysten commented on .changeset/tender-dots-obey.md:
```diff
@@ -0,0 +1,5 @@
+'@mysten/zklogin': minor
+---
+
+add support for verify zklogin personal msg
```
could update this to just say verify signatures since you added transaction block verification too
